eve/main.yml

```diff
@@ -1553,3 +1556,295 @@ stages:
+          haltOnFailure: true
+      - ShellCommand:
+          name: Copy example-solution CRs to Bootstrap node
+          env:
+            SSH_CONFIG: >-
+              eve/workers/openstack-multiple-nodes/terraform/ssh_config
+            HOST_CR_PATH: "examples/metalk8s-solution-example/example-operator/deploy/crds/examplesolution_v1alpha1_versionserver_cr.yaml"
+            BOOTSTRAP_CR_PATH: "/tmp/examplesolution_v1alpha1_versionserver_cr.yaml"
+          command: >
+            scp -F $SSH_CONFIG "${HOST_CR_PATH}" bootstrap:"${BOOTSTRAP_CR_PATH}"
+          haltOnFailure: true
+      - ShellCommand:
+          name: Apply example-solution CRs on Bootstrap node
+          command: >
+            ssh -F ssh_config bootstrap
+            sudo kubectl --kubeconfig /etc/kubernetes/admin.conf apply -f
+            /tmp/examplesolution_v1alpha1_versionserver_cr.yaml
+          workdir: build/eve/workers/openstack-multiple-nodes/terraform/
+          haltOnFailure: true
+      - ShellCommand:
+          name: Verify example-solution-operator Deployment
+          env:
+            KUBECONFIG: "/etc/kubernetes/admin.conf"
+          command: |
+            ssh -F ssh_config bootstrap --
+            bash -c '
+            ATTEMPTS=0
+            declare -a GET_REPLICAS=(
+                sudo kubectl --kubeconfig $KUBECONFIG get deployments
+                example-solution-operator -n example-environment
+                -o custom-columns=:status.availableReplicas
+            )
+            REPLICAS=$("${GET_REPLICAS[@]}")
+            until [ $REPLICAS -eq 1 ]; do
+              ((ATTEMPTS++))
+              if [ $ATTEMPTS -eq 10 ]; then
+                echo "Failed to obtain Replicas after ${ATTEMPTS} attempts" >&2
+                exit 1
+              fi
+              sleep 3
+              REPLICAS=$("${GET_REPLICAS[@]}")
+            done'
+          workdir: build/eve/workers/openstack-multiple-nodes/terraform/
+          haltOnFailure: true
+      - ShellCommand:
+          name: Verify exampleversions Deployment
+          env:
+            KUBECONFIG: "/etc/kubernetes/admin.conf"
+          command: |
+            ssh -F ssh_config bootstrap --
+            bash -c '
+            ATTEMPTS=0
+            declare -a GET_REPLICAS=(
+                sudo kubectl --kubeconfig $KUBECONFIG get deployments
+                example-versionserver -n example-environment
+                -o custom-columns=:status.availableReplicas
+            )
+            REPLICAS=$("${GET_REPLICAS[@]}")
+            until [ $REPLICAS -eq 3 ]; do
+              ((ATTEMPTS++))
+              if [ $ATTEMPTS -eq 10 ]; then
+                echo "Failed to obtain Replicas after ${ATTEMPTS} attempts" >&2
+                exit 1
+              fi
+              sleep 3
+              REPLICAS=$("${GET_REPLICAS[@]}")
+            done'
+          workdir: build/eve/workers/openstack-multiple-nodes/terraform/
+          haltOnFailure: true
```

Comment:
Not tested, but something like that should be better (less code duplication):
```suggestion
      - ShellCommand: &check_k8s_resource_column_ssh
          name: Verify example-solution-operator Deployment
          env: &_env_check_k8s_resource_column_ssh
            KUBECONFIG: "/etc/kubernetes/admin.conf"
            NAMESPACE: "example-environment"
            RESOURCE_TYPE: "deployment"
            RESOURCE_NAME: "example-solution-operator"
            RESOURCE_COLUMN: ":status.availableReplicas"
            RESOURCE_VALUE: "1"
            RETRIES: "10"
            SLEEP: "3"
          command: |
            ssh -F ssh_config bootstrap --
            bash -c '
            ATTEMPTS=0
            declare -a GET_COLUMN_VALUE=(
                sudo kubectl --kubeconfig "$KUBECONFIG" get "$RESOURCE_TYPE"
                "$RESOURCE_NAME" -n "$NAMESPACE"
                -o custom-columns="$RESOURCE_COLUMN"
            )
            CURRENT_VALUE=$("${GET_COLUMN_VALUE[@]}")
            until [ "$CURRENT_VALUE" -eq "$WANTED_VALUE" ]; do
              (( ATTEMPTS++ ))
              if [ "$ATTEMPTS" -eq "$RETRIES" ]; then
                echo "$RESOURCE_COLUMN still not equal to $RESOURCE_VALUE after $ATTEMPTS attempts" >&2
                exit 1
              fi
              sleep "$SLEEP"
              CURRENT_VALUE=$("${GET_COLUMN_VALUE[@]}")
            done'
          workdir: build/eve/workers/openstack-multiple-nodes/terraform/
          haltOnFailure: true
      - ShellCommand:
          <<: *check_k8s_resource_column_ssh
          name: Verify example-versionserver Deployment
          env:
            <<: *_env_check_k8s_resource_column_ssh
            RESOURCE_NAME: "example-versionserver"
            RESOURCE_VALUE: "3"
```

Reply:
Anyway, I'm not even sure the first approach or this one can work, using simple quotes will prevent environment variables from being expanded in the string passed to `bash -c`... and if you use double quotes, some variables will be expanded before the script got executed.
Maybe it would be better to put this in a dedicated `.sh`.

Reply:
Hmm, i would prefer doing it without a dedicated shell script

Reply:
You can, you just need to use double quotes around the whole bash script and escape things that must not be expanded the first time.